modes.go

```diff
@@ -343,48 +389,79 @@ func DoBatchedWrites(session *gocql.Session, threadResult *results.TestThreadRes
 			batch.Query(request, currentPk, ck, value)
 		}
 
-		requestStart := time.Now()
-		err := session.ExecuteBatch(batch)
-		requestEnd := time.Now()
-		if err != nil {
-			return err, time.Duration(0)
+		query_str := "BATCH >>>" + request + "; ck = " + fmt.Sprint(cks) + "; v_sizes = " + fmt.Sprint(values_sizes)
+		query_str = strings.Replace(query_str, "(?", "("+strconv.FormatInt(currentPk, 10), 1)
+		current_attempts := 0
+		for {
+			requestStart := time.Now()
+			err := session.ExecuteBatch(batch)
+			requestEnd := time.Now()
+
+			if err == nil {
+				rb.IncOps()
+				rb.AddRows(batchSize)
+				latency := requestEnd.Sub(requestStart)
+				return nil, latency
+			}
+			if retryHandler == "sb" {
+				err = handle_sb_retry_error(query_str, err, current_attempts)
+			}
+			if err != nil {
+				return err, time.Duration(0)
+			}
+			current_attempts++
 		}
-
-		rb.IncOps()
-		rb.AddRows(batchSize)
-
-		latency := requestEnd.Sub(requestStart)
-		return nil, latency
 	})
 }
 
 func DoCounterUpdates(session *gocql.Session, threadResult *results.TestThreadResult, workload WorkloadGenerator, rateLimiter RateLimiter) {
 	RunTest(threadResult, workload, rateLimiter, func(rb *results.TestThreadResult) (error, time.Duration) {
-		query := session.Query("UPDATE " + keyspaceName + "." + counterTableName +
-			" SET c1 = c1 + ?, c2 = c2 + ?, c3 = c3 + ?, c4 = c4 + ?, c5 = c5 + ? WHERE pk = ? AND ck = ?")
+		request := "UPDATE " + keyspaceName + "." + counterTableName +
+			" SET c1 = c1 + ?, c2 = c2 + ?, c3 = c3 + ?, c4 = c4 + ?, c5 = c5 + ? WHERE pk = ? AND ck = ?"
+		query := session.Query(request)
 		pk := workload.NextPartitionKey()
 		ck := workload.NextClusteringKey()
+		current_attempts := 0
+		query_str := request
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck, 10), 1)
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck+1, 10), 1)
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck+2, 10), 1)
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck+3, 10), 1)
+		query_str = strings.Replace(query_str, "+ ?", "+ "+strconv.FormatInt(ck+4, 10), 1)
+		query_str = strings.Replace(query_str, "pk = ?", "pk = "+strconv.FormatInt(pk, 10), 1)
+		query_str = strings.Replace(query_str, "ck = ?", "ck = "+strconv.FormatInt(ck, 10), 1)
 		bound := query.Bind(ck, ck+1, ck+2, ck+3, ck+4, pk, ck)
+		for {
+			requestStart := time.Now()
+			err := bound.Exec()
+			requestEnd := time.Now()
 
-		requestStart := time.Now()
-		err := bound.Exec()
-		requestEnd := time.Now()
-		if err != nil {
-			return err, time.Duration(0)
+			if err == nil {
+				rb.IncOps()
+				rb.IncRows()
+				latency := requestEnd.Sub(requestStart)
+				return nil, latency
+			}
+			if retryHandler == "sb" {
+				err = handle_sb_retry_error(query_str, err, current_attempts)
+			}
+			if err != nil {
+				return err, time.Duration(0)
+			}
+			current_attempts++
```

Comment:
Nit: this pattern appears in multiple functions. You can probably abstract it away:

```go
func runWithRetries(op func() error) {
    for {
        requestStart := time.Now()
        err := op()
        requestEnd := time.Now()

        // ...
    }
}
```

Reply:
Right, `op` would be different in all three cases. For `DoWrites` it will probably be just:

```go
runWithRetries(func() error {
    return bound.Exec()
})
```

For `DoScanTable`:

```go
runWithRetries(func() error {
    bound := query.Bind(currentRange.Start, currentRange.End)
    iter := bound.Iter()
    for iter.Scan(nil, nil, nil) {
        rb.IncRows()
    }
    return iter.Close() // <- in the original code, `requestEnd := time.Now()` was before this line, but I think it's perfectly fine to reorder
}
```

For `DoReadsFromTable` it will probably be quite long, but it's copy paste from the existing code.

That said - I'm making those up on the spot, maybe I'm missing something. It would be nice to achieve this deduplication, but if it's problematic then this refactor is not super important and there is no need to dwell on it too much.

Reply:
Still sounds as `too much` for the goal of this PR.